File: src/module_mesh_utils.f90
module module_mesh_utils

  use iso_fortran_env , only : dp => REAL64

  implicit none

  interface distance
     module procedure distanceAB
  end interface distance

contains

  !===================================================================!
  ! Cross product for area computations
  ! ===================================================================!

  pure subroutine cross_product(a, b, pdt)

    real(dp), intent(in)  :: a(3), b(3)
    real(dp), intent(out) :: pdt(3)

    ! use skew form and generalize to n-dimensions?
    pdt(1) = a(2) * b(3) - a(3) * b(2)
    pdt(2) = a(3) * b(1) - a(1) * b(3)
    pdt(3) = a(1) * b(2) - a(2) * b(1)

  end subroutine cross_product

  !===================================================================!
  ! Geometric distance between two points
  !===================================================================!

  pure real(dp) function distanceAB(x, y)

    real(dp), intent(in)  :: X(:), y(:) ! [[x,y,z], [1:2]]

    distanceAB = sqrt(sum((x-y)**2))

  end function distanceAB

  !===================================================================!
  ! Index of a target value if present in the array
  !===================================================================!

  pure type(integer) function find(array, target_value)

    integer, intent(in) :: array(:)
    integer, intent(in) :: target_value
    integer :: i, nentries

    nentries = size(array, dim=1)

    do i = 1, nentries
       if (array(i) .eq. target_value) then
          find = i
          return
       endif
    end do

    find = -1

  end function find

  !===================================================================!
  ! Checks if the first smaller array is a subset of the second larger
  ! array
  !===================================================================!

  pure type(logical) function is_subset(small, big)

    integer, intent(in) :: small(:)
    integer, intent(in) :: big(:)

    integer, allocatable :: sub(:)
    integer, allocatable :: set(:)

    integer :: lensub, i

    is_subset = .false.

    if (size(small) .gt. size(big)) return

    ! Create local copy of arrays
    allocate(sub, source = small)
    allocate(set, source = big)

    ! Sort two arrays
    call isort(sub)
    call isort(set)
    lensub = size(sub)

    ! Check if all entries are equal upto the length of the smallest
    ! array
    is_subset = .true.
    do i = 1, lensub
       if (any(set .eq. sub(i)) .eqv. .false.) then
          is_subset = .false.
          exit
       end if
    end do

    deallocate(sub,set)

  end function is_subset

  !===================================================================!
  ! Sort an integer array ! move elsewhere?
  !===================================================================!

  pure subroutine isort(array)

    integer, intent(inout) :: array(:)
    integer :: temp , j , k
    integer :: n

    n = ubound(array,1)

    do j = 1 , n
       do k = j + 1 , n
          if(array(j) > array(k)) then
             temp     = array(k)
             array(k) = array(j)
             array(j) = temp
          end if
       end do
    end do

  end subroutine isort

  ! generalize the name to return the number of lower dimensional entities
  pure elemental integer function elem_type_face_count(cell_type) result (num_faces)

    integer, intent(in) :: cell_type

    select case (cell_type)
    case (1)
       ! 2-node line.
       num_faces = 2 ! vertex
    case (2)
       ! 3-node triangle
       num_faces = 3  ! 3 edges
    case (3)
       ! 4-node quadrangle
       num_faces = 4 ! 4 edges
    case (4)
       ! 4-node tetrahedron
       num_faces = 4
    case (5)
       ! 8-node hexahedron
       num_faces = 6
    case (6)
       !  6-node prism
       num_faces = 5
    case(7)
       ! 5-node prism (pyramid)
       num_faces = 5
    case default
       num_faces = 0
    end select

  end function elem_type_face_count

  ! Spatial dimension of a gmsh element type (point=0, line=1,
  ! tri/quad=2, tet/hex/prism/pyramid=3). Used to classify elements into
  ! cells / faces / edges by dimension rather than by type.
  pure elemental integer function elem_type_dimension(elem_type) result (dim)

    integer, intent(in) :: elem_type

    select case (elem_type)
    case (15)
       ! 1-node point
       dim = 0
    case (1)
       ! 2-node line
       dim = 1
    case (2:3)
       ! triangle, quadrangle
       dim = 2
    case (4:7)
       ! tet, hex, prism, pyramid
       dim = 3
    case default
       dim = -1
    end select

  end function elem_type_dimension

  ! generalize the name to return the number of lower dimensional entities
  pure elemental integer function elem_type_vertex_count(elem_type) result (num_vertices)

    integer, intent(in) :: elem_type

    select case (elem_type)
    case (1)
       ! 2-node line.
       num_vertices = 2
    case (2)
       ! 3-node triangle
       num_vertices = 3
    case (3)
       ! 4-node quadrangle
       num_vertices = 4
    case (4)
       ! 4-node tetrahedron
       num_vertices = 4
    case (5)
       ! 8-node hexahedron
       num_vertices = 8
    case (6)
       !  6-node prism
       num_vertices = 6
    case(7)
       ! 5-node prism (pyramid)
       num_vertices = 5
    case default
       num_vertices = 0
    end select

  end function elem_type_vertex_count

  impure subroutine order_face_vertices(cell_type, cell_vertices, face_vertices_unordered)

    integer, intent(in)    :: cell_type
    integer, intent(in)    :: cell_vertices(:)
    integer, intent(inout) :: face_vertices_unordered(:)

    integer, allocatable :: face_vertices(:,:)
    integer :: num_face_vertices, num_cell_faces
    integer :: iface, ivertex
    integer :: match_count

    select case (cell_type)
    case (2)
       ! 3-node triangle (2d cell) - faces are its 3 edges
       num_face_vertices = 2
       num_cell_faces = 3

       if (num_face_vertices .ne. size(face_vertices_unordered)) &
            & error stop "inconstent vertices"

       allocate(face_vertices(num_face_vertices, num_cell_faces))
       face_vertices = 0

       face_vertices(:,1) = [cell_vertices(1), cell_vertices(2)]
       face_vertices(:,2) = [cell_vertices(2), cell_vertices(3)]
       face_vertices(:,3) = [cell_vertices(3), cell_vertices(1)]

    case (3)
       ! 4-node quadrangle (2d cell) - faces are its 4 edges
       num_face_vertices = 2
       num_cell_faces = 4

       if (num_face_vertices .ne. size(face_vertices_unordered)) &
            & error stop "inconstent vertices"

       allocate(face_vertices(num_face_vertices, num_cell_faces))
       face_vertices = 0

       face_vertices(:,1) = [cell_vertices(1), cell_vertices(2)]
       face_vertices(:,2) = [cell_vertices(2), cell_vertices(3)]
       face_vertices(:,3) = [cell_vertices(3), cell_vertices(4)]
       face_vertices(:,4) = [cell_vertices(4), cell_vertices(1)]

    case (4)
       ! 4-node tetrahedron
       num_face_vertices = 3
       num_cell_faces = 4

       if (num_face_vertices .ne. size(face_vertices_unordered)) &
            & error stop "inconstent vertices"

       allocate(face_vertices(num_face_vertices, num_cell_faces))
       face_vertices = 0

       ! normals must point outward
       face_vertices(:,1) = [cell_vertices(1), cell_vertices(2), cell_vertices(3)]
       face_vertices(:,2) = [cell_vertices(3), cell_vertices(2), cell_vertices(4)]
       face_vertices(:,3) = [cell_vertices(1), cell_vertices(3), cell_vertices(4)]
       face_vertices(:,4) = [cell_vertices(1), cell_vertices(4), cell_vertices(2)]

    case (5)
       ! 8-node hexahedron
       num_face_vertices = 4
       num_cell_faces = 6

       if (num_face_vertices .ne. size(face_vertices_unordered)) &
            & error stop "inconstent vertices"

       allocate(face_vertices(num_face_vertices, num_cell_faces))
       face_vertices = 0

       face_vertices(:,1) = [cell_vertices(1), cell_vertices(5), cell_vertices(8), cell_vertices(4)]
       face_vertices(:,2) = [cell_vertices(2), cell_vertices(3), cell_vertices(7), cell_vertices(6)]
       face_vertices(:,3) = [cell_vertices(5), cell_vertices(6), cell_vertices(7), cell_vertices(8)]
       face_vertices(:,4) = [cell_vertices(1), cell_vertices(4), cell_vertices(3), cell_vertices(2)]
       face_vertices(:,5) = [cell_vertices(7), cell_vertices(3), cell_vertices(4), cell_vertices(8)]
       face_vertices(:,6) = [cell_vertices(1), cell_vertices(2), cell_vertices(6), cell_vertices(5)]

    case (6)
       !  6-node prism (5 faces)
       num_cell_faces = 5
       num_face_vertices = 4 ! take the max

       allocate(face_vertices(num_face_vertices, num_cell_faces))
       face_vertices = 0

       face_vertices(:,1) = [cell_vertices(1), cell_vertices(4), cell_vertices(5), cell_vertices(6)]
       face_vertices(:,2) = [cell_vertices(1), cell_vertices(3), cell_vertices(6), cell_vertices(4)]
       face_vertices(:,3) = [cell_vertices(2), cell_vertices(5), cell_vertices(6), cell_vertices(3)]
       face_vertices(:,4) = [cell_vertices(1), cell_vertices(2), cell_vertices(3)]
       face_vertices(:,5) = [cell_vertices(4), cell_vertices(6), cell_vertices(5)]

    case(7)
       ! 5-node prism (pyramid) 5 faces
       num_cell_faces = 5
       num_face_vertices = 4 ! take the max

       allocate(face_vertices(num_face_vertices, num_cell_faces))
       face_vertices = 0

       face_vertices(:,1) = [cell_vertices(1), cell_vertices(2), cell_vertices(3), cell_vertices(4)]
       face_vertices(:,2) = [cell_vertices(1), cell_vertices(2), cell_vertices(5)]
       face_vertices(:,3) = [cell_vertices(2), cell_vertices(3), cell_vertices(5)]
       face_vertices(:,4) = [cell_vertices(3), cell_vertices(4), cell_vertices(5)]
       face_vertices(:,5) = [cell_vertices(1), cell_vertices(5), cell_vertices(4)]

    case default
       print *, "unknown type"
       error stop
    end select

    ! compare the input with face 1,2,3,4 and see which has the
    ! matching count equal to input vertex count

    match_face: do iface = 1, num_cell_faces

       match_count = 0
       match_vertex: do ivertex = 1, num_face_vertices
          if (any (face_vertices_unordered(:) .eq. face_vertices(ivertex,iface) ) .eqv. .true.) then
             ! exit match_vertex ! skip and go to next face
             ! else
             match_count = match_count + 1
          end if
       end do match_vertex

       ! found the face with correct ordering of vertices
       if (match_count .eq. num_face_vertices) then
          face_vertices_unordered = face_vertices(:, iface)
          exit match_face
       end if

    end do match_face

    if (allocated(face_vertices)) deallocate(face_vertices)

  end subroutine order_face_vertices

end module module_mesh_utils

File: ufvm/module_mesh_utils.py
#=====================================================================#
# module_mesh_utils - free functions for mesh connectivity / geometry
#
# Index convention for the port: connectivity arrays store 1-based IDs
# as *values* (as the Fortran does), while array *positions* are 0-based.
# An ID is converted to a position with (id - 1); a 0-based loop position
# is converted back to a stored ID with (position + 1).
#=====================================================================#

import numpy as np


#===================================================================#
# Cross product for area computations
#===================================================================#
def cross_product(a, b, pdt):
    # use skew form and generalize to n-dimensions?
    pdt[0] = a[1] * b[2] - a[2] * b[1]
    pdt[1] = a[2] * b[0] - a[0] * b[2]
    pdt[2] = a[0] * b[1] - a[1] * b[0]


#===================================================================#
# Geometric distance between two points
#===================================================================#
def distanceX(X):
    # X(:,:) -> [[x,y,z], [1:2]]
    return np.sqrt(np.sum((X[:, 0] - X[:, 1]) ** 2))


def distanceAB(x, y):
    return np.sqrt(np.sum((np.asarray(x) - np.asarray(y)) ** 2))


def distance(*args):
    """Generic interface `distance` dispatching on argument count."""
    if len(args) == 1:
        return distanceX(args[0])
    else:
        return distanceAB(args[0], args[1])


#===================================================================#
# Invert a map. The keys of 'map' are values of 'inverse'. The
# values of 'map' are the keys of 'inverse'
#===================================================================#
def reverse_map(map, num_map_vals):
    map = np.asarray(map)
    num_map_vals = np.asarray(num_map_vals)

    if num_map_vals.size == 0:
        return None, None

    # Forward mapping size
    nkeysin = map.shape[1]
    nvalsin = map.shape[0]

    # Nothing to do (probably empty map)!
    if nkeysin == 0:
        return None, None

    # Find the maximum size of the inverse map values
    nkeysout = int(map.max())   # dangerous
    num_inverse_vals = np.zeros(nkeysout, dtype=int)
    for i in range(nkeysin):
        for j in range(num_map_vals[i]):
            value = map[j, i]
            num_inverse_vals[value - 1] = num_inverse_vals[value - 1] + 1
    nvalsout = int(num_inverse_vals.max())

    # Allocate the inverse map based on determined sizes
    inverse = np.zeros((nvalsout, nkeysout), dtype=int)

    # Point into the next available slot for each inverse_key
    ptr = np.zeros(nkeysout, dtype=int)
    for i in range(nkeysin):
        for j in range(num_map_vals[i]):
            value = map[j, i]
            ptr[value - 1] = ptr[value - 1] + 1
            inverse[ptr[value - 1] - 1, value - 1] = i + 1   # store key as 1-based id

    return inverse, num_inverse_vals


#===================================================================#
# Find the intersection of two arrays (move elsewhere)?
# Writes the first common element into c (in place), as in Fortran.
#===================================================================#
def intersection(a, b, c):
    sizea = np.asarray(a).size
    sizeb = np.asarray(b).size

    ctr = 0
    for i in range(sizea):
        for j in range(sizeb):
            # Copy entry to new list if equal
            if a[i] == b[j]:
                ctr = ctr + 1
                c[ctr - 1] = a[i]
                return


#===================================================================#
# Find if a target value is present in the array (move else where)?
# Returns the 1-based index of the value (or -1), matching Fortran.
#===================================================================#
def find(array, target_value):
    nentries = np.asarray(array).shape[0]

    for i in range(nentries):
        if array[i] == target_value:
            return i + 1

    return -1


#===================================================================#
# Checks if the first argument is a subset of the second argument
#===================================================================#
def is_subset(small, big):
    small = np.asarray(small)
    big = np.asarray(big)

    if small.size > big.size:
        return False

    # Create local copy of arrays
    sub = np.array(small)
    set_ = np.array(big)

    # Sort two arrays
    isort(sub)
    isort(set_)
    lensub = sub.size

    # Check if all entries are equal upto the length of the smallest array
    for i in range(lensub):
        if not (set_ == sub[i]).any():
            return False
    return True


#===================================================================#
# Sort an integer array ! move elsewhere?
#===================================================================#
def isort(array):
    n = array.shape[0]
    for j in range(n):
        for k in range(j + 1, n):
            if array[j] > array[k]:
                temp = array[k]
                array[k] = array[j]
                array[j] = temp


#===================================================================#
# Forms the cell faces from a pair of vertices belonging to cell.
#===================================================================#
def get_cell_faces(cell_vertices, vertex_faces, num_vertex_faces):
    cell_vertices = np.asarray(cell_vertices)
    vertex_faces = np.asarray(vertex_faces)
    num_vertex_faces = np.asarray(num_vertex_faces)

    nvertices = cell_vertices.shape[0]
    nfaces = nvertices
    ncells = cell_vertices.shape[1]

    # find how many faces are there based on nodes
    num_cell_faces = np.zeros(ncells, dtype=int)
    for icell in range(ncells):
        ctr = 0
        for iface in range(nvertices):
            if cell_vertices[iface, icell] != 0:
                ctr = ctr + 1
        num_cell_faces[icell] = ctr

    # Cell to face cell_vertices
    cell_faces = np.zeros((int(num_cell_faces.max()), ncells), dtype=int)
    for icell in range(ncells):
        face_ptr = 0
        for iface in range(num_cell_faces[icell]):
            # Get the first two vertices
            if iface == num_cell_faces[icell] - 1:
                v1 = cell_vertices[iface, icell]
                v2 = cell_vertices[0, icell]
            else:
                v1 = cell_vertices[iface, icell]
                v2 = cell_vertices[iface + 1, icell]

            face_ptr = face_ptr + 1
            # v1, v2 are 1-based vertex ids -> index with (v - 1)
            intersection(
                vertex_faces[0:num_vertex_faces[v2 - 1], v2 - 1],
                vertex_faces[0:num_vertex_faces[v1 - 1], v1 - 1],
                cell_faces[face_ptr - 1:face_ptr, icell])

    return cell_faces, num_cell_faces


#===================================================================#
# Determine if the face is a boundary face based on how many
# neighbouring cells it has.
#===================================================================#
def get_boundary_faces(num_face_cells):
    num_face_cells = np.asarray(num_face_cells)
    nfaces = num_face_cells.shape[0]

    # Boundary faces are the faces corresponding to just one cell
    nbfaces = 0
    for iface in range(nfaces):
        if num_face_cells[iface] == 1:
            nbfaces = nbfaces + 1

    boundary_faces = np.zeros(nbfaces, dtype=int)
    ctr = 0
    for iface in range(nfaces):
        if num_face_cells[iface] == 1:
            ctr = ctr + 1
            boundary_faces[ctr - 1] = iface + 1   # store 1-based face id

    return boundary_faces
